File: desktop/Protocol/HeartbeatPayload.cs
namespace CTRL.Desktop.Protocol;

public sealed record HeartbeatPayload(ulong ClientSendTime);

public static class HeartbeatPayloadCodec
{
    public static byte[] Encode(HeartbeatPayload payload)
    {
        var writer = new PayloadWriter();
        writer.WriteUInt64(payload.ClientSendTime);
        return writer.ToArray();
    }

    public static HeartbeatPayload Decode(byte[] payload)
    {
        var reader = new PayloadReader(payload);
        var clientSendTime = reader.ReadUInt64();
        reader.ExpectEnd();
        return new HeartbeatPayload(clientSendTime);
    }
}

File: mobile/lib/protocol/heartbeat_payload.dart
import 'dart:typed_data';

import 'payload_codec.dart';

class HeartbeatPayload {
  const HeartbeatPayload({required this.clientSendTime});

  final int clientSendTime;

  @override
  bool operator ==(Object other) =>
      other is HeartbeatPayload && clientSendTime == other.clientSendTime;

  @override
  int get hashCode => clientSendTime.hashCode;
}

class HeartbeatPayloadCodec {
  static Uint8List encode(HeartbeatPayload payload) {
    final writer = PayloadWriter()..writeUInt64(payload.clientSendTime);
    return writer.toBytes();
  }

  static HeartbeatPayload decode(List<int> bytes) {
    final reader = PayloadReader(bytes);
    final clientSendTime = reader.readUInt64();
    reader.expectEnd();
    return HeartbeatPayload(clientSendTime: clientSendTime);
  }
}
